File: packages/installer-server/src/lib/steps/installOpenclaw.ts
import { execa } from 'execa';
import type { InstallerStep } from './index.js';

async function detectPackageManager(): Promise<'pnpm' | 'bun' | 'npm'> {
  for (const candidate of ['pnpm', 'bun'] as const) {
    try {
      await execa(candidate, ['--version'], { stdio: 'ignore' });
      return candidate;
    } catch {
      // try next
    }
  }
  return 'npm';
}

export const installOpenclawStep: InstallerStep = {
  id: 'install-openclaw',
  label: 'Install OpenClaw CLI globally',
  shouldRun: () => true,
  async run({ bus }) {
    const pm = await detectPackageManager();
    bus.log(installOpenclawStep.id, 'info', `Using package manager: ${pm}`);

    const args =
      pm === 'pnpm'
        ? ['add', '-g', '@openclaw/cli@latest']
        : pm === 'bun'
          ? ['add', '-g', '@openclaw/cli@latest']
          : ['install', '-g', '@openclaw/cli@latest'];

    const sub = execa(pm, args, { all: true });
    sub.all?.on('data', (chunk: Buffer) => {
      const text = chunk.toString().trimEnd();
      if (text) bus.log(installOpenclawStep.id, 'info', text);
    });
    await sub;
  },
};

File: internal/installer/steps/install_openclaw.go
package steps

import (
	"context"

	"github.com/SolsticeSauer/solsclaw_beta/internal/installer"
)

type InstallOpenclaw struct{}

func (InstallOpenclaw) ID() string                              { return "install-openclaw" }
func (InstallOpenclaw) Label() string                           { return "Install OpenClaw CLI globally" }
func (InstallOpenclaw) ShouldRun(_ installer.StepContext) bool  { return true }

func (s InstallOpenclaw) Run(ctx context.Context, sc installer.StepContext) error {
	pm, args := pickPackageManager()
	sc.Bus.Log(s.ID(), "info", "Using package manager: "+pm)
	return streamCommand(ctx, sc, s.ID(), pm, args...)
}

// pickPackageManager prefers pnpm > bun > npm because OpenClaw upstream
// recommends pnpm. We don't try to auto-install Node here — the bootstrap
// script handles that prerequisite.
func pickPackageManager() (string, []string) {
	switch {
	case commandExists("pnpm"):
		return "pnpm", []string{"add", "-g", "openclaw@latest"}
	case commandExists("bun"):
		return "bun", []string{"add", "-g", "openclaw@latest"}
	default:
		return "npm", []string{"install", "-g", "openclaw@latest"}
	}
}
